backend/src/main/java/com/javahelp/backend/data/DynamoDBSurveyResponseStore.java

```diff
@@ -209,43 +212,26 @@ public SurveyResponse readByUserAndSurvey(String userID, String surveyID) {
      * with the specific constraints.
      */
     @Override
-    public Map<String, SurveyResponse> readByConstraint(Map<String, Set<String>> constraint) {
+    public Map<String, SurveyResponse> readByConstraint(IConstraint constraint) {
         Map<String, AttributeValue> expressionAttributeValues = new HashMap<>();
+        expressionAttributeValues.put(":provider_survey_val", new AttributeValue().withBOOL(true));
 
-        int n_questions = 0;
-        int n_answers = 0;
+        int n_attr = 1;
         StringBuilder conditionBuilder = new StringBuilder();
-        for (String question : constraint.keySet()) {
-            String formattedQuestion = String.join("_", question.split(" "));
-            n_questions = n_questions + 1;
-            Set<String> answers = constraint.get(question);
-
-            Set<String> expressions = new HashSet<>();
-            for (String answer: answers) {
-                String keyString = String.format(":answer_%s", n_answers);
-                expressions.add(keyString);
-                expressionAttributeValues.put(keyString, new AttributeValue().withN(answer));
-                n_answers = n_answers + 1;
-            }
-            String attrKeyString = String.format("responses.%s", formattedQuestion);
-            String attrKeySetString = String.format("(%s)", String.join(", ", expressions));
-            conditionBuilder.append("(");
-            conditionBuilder.append(attrKeyString);
-            conditionBuilder.append(" IN ");
-            conditionBuilder.append(attrKeySetString);
-            conditionBuilder.append(")");
-            if (n_questions < constraint.size()) {
-                conditionBuilder.append(" AND ");
-            }
+        for (String attr : constraint.getConstraints()) {
+            String keyString = String.format(":attr_%s", n_attr);
+            expressionAttributeValues.put(keyString, new AttributeValue().withS(attr));
+
+            conditionBuilder.append(String.format("(contains(attributes, %s))", keyString));
+            conditionBuilder.append(" AND ");
+
+            n_attr = n_attr + 1;
         }
```

Comment:
Good to see you used design pattern here to make the coding more clean and easy to read.